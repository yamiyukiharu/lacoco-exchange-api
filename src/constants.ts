export enum Token {
  BTC = 'BTC',
  ETH = 'ETH',
  DFI = 'DFI',
  USDT = 'USDT',
  DOGE = 'DOGE',
  BNB = 'BNB',
}
