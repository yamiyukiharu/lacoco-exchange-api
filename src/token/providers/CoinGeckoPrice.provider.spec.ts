import { HttpService } from '@nestjs/axios';
import { Test, TestingModule } from '@nestjs/testing';
import { CoinGeckoPriceProvider } from './CoinGeckoPrice.provider';
import { Token } from '../../constants';

describe('CoinGeckPriceProvider', () => {
  let provider: CoinGeckoPriceProvider;
  const httpService = {
    axiosRef: {
      get: jest.fn(),
    },
  };

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        { provide: HttpService, useValue: httpService },
        CoinGeckoPriceProvider,
      ],
    }).compile();

    provider = module.get<CoinGeckoPriceProvider>(CoinGeckoPriceProvider);
  });

  it('can get token prices', async () => {
    const res = {
      bitcoin: { usd: 60000 },
      ethereum: { usd: 3000 },
    };
    const fn = httpService.axiosRef.get;
    fn.mockResolvedValue({ data: res });

    const prices = await provider.getPrices([Token.BTC, Token.ETH]);

    expect(prices[Token.BTC]).toStrictEqual(res.bitcoin.usd);
    expect(prices[Token.ETH]).toStrictEqual(res.ethereum.usd);
    expect(fn).toBeCalledWith('https://api.coingecko.com/api/v3/simple/price', {
      params: {
        ids: 'bitcoin,ethereum',
        vs_currencies: 'usd',
      },
    });
  });

  it('throws error if token is not found', async () => {
    const res = {
      bitcoin: { usd: 60000 },
      ethereum: { usd: 3000 },
    };
    const fn = httpService.axiosRef.get;
    fn.mockResolvedValue({ data: res });

    await expect(
      provider.getPrices([Token.BTC, Token.ETH, Token.BNB]),
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
