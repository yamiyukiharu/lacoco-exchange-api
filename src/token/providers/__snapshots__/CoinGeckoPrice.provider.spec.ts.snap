// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoinGeckPriceProvider throws error if token is not found 1`] = `"Invalid tokens: BNB"`;
